### PR TITLE
fix(curriculum): Added introduction to meta tag

### LINKED ISSUE #47118: CatPhotoApp Step 68: Issues with `meta` element
### Describe the Issue

I believe this is the first step to introduce the `meta` element. Two issues with this step:
- [ ] The description does not explain that it is a self-closing tag
- [ ] The tests accept a closing tag

### Affected Page

https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-html-by-building-a-cat-photo-app/step-68

### Your code

You can pass this step with the following:

```
<meta charset="UTF-8"></meta>
```

### Expected behavior

Since this is the first time the `meta` element is used we should explain in the description that it is a self-closing tag. I'm assuming that we are not expecting the user to look up this information themselves.

Also, we should fix the test so that it does not accept a closing `</meta>` tag.

### Screenshots

_No response_

### System

N/A

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by hanswang123456**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #47118

### CODE REVIEW

**Comment by bbsmooth** [comment]
I could be wrong about this, but I seem to recall someone mentioning that "we" should be avoided in the description.

> The `meta` element is used to add additional information to the page. Like the `img` element, the `meta` element uses a self-closing tag. To help the browser display the text on the page properly, add a `meta` element as a child of the `head` element and set its `charset` attribute to `UTF-8`.

**Comment by hanswang123456** [comment]
@bbsmooth I've changed the phrasing up and have a draft regex. I still need to test the regex and see if I have to do a full regex or not.

**Comment by bbsmooth** [comment]
I appreciate the effort, but I still think the description needs work. I think it's too repetitive and verbose. And we aren't requiring forward slashes at the end self-closing tags any more so I would not include one in the description. And according to MDN, the only valid value for `charset` is `UTF-8`, so it's not just the most common encoding scheme, it's the only one. And the W3C will throw an error if you use any other encoding scheme. In fact, I believe you don't need to set the `charset` at all as all browsers will automatically default to `UTF-8` and the W3C validator doesn't complain if you leave it out. So perhaps this step ought to just be removed?

**Comment by hanswang123456** [comment]
I've simplified the description but haven't removed the charset step yet. should we wait for some other input?

**Comment by bbsmooth** [comment]
Ya, removing a step in a course is not something the two of us are going to decide :smile: There may be other legit reasons for keeping this step.

**Comment by hanswang123456** [comment]
Yea, I think the placement of the step should be closer to the start. Hopefully we get some more opinions on it.

**Comment by naomi-lgbt** [comment]
Hmm, it looks like the `charset` used to take valid options like `ascii` and `utf-16`, but doesn't in HTML5?

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md [hunk 1]
```diff
@@ -6,15 +6,19 @@ dashedName: step-68
 ---
 
 # --description--
+In this step, a `meta` element will be added to the `head`. This self-closing element communicates metadata to the browser and can be written as `<meta>`. In this step, you will add a `charset` attribute which describes the character encoding to the browser.
```
Comment (bbsmooth) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:9]: I don't think we need to tell them that we are adding a meta element to the head in the first sentence because we are going to tell them to do that in the actual instructions. Same with telling them that they are going to add a charset attribute just before the instructions tell them to add a charset attribute.

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md [hunk 2]
```diff
@@ -6,15 +6,19 @@ dashedName: step-68
 ---
 
 # --description--
```
Comment (Sboonny) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:8]: I don't know how you bypass enlist errors, but this should have given you an error that you are missing an extra space 

```suggestion
# --description--

```

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md [hunk 3]
```diff
@@ -6,15 +6,19 @@ dashedName: step-68
...
+To communicate meta data to the browser, a `<meta>` tag is used. An example of metadata is `charset`, which describes the character encoding of the document.
```
Comment (Sboonny) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:10]:
```suggestion
You can set the browser behavior by using `meta` element. A famous use case is telling the browser to parse the markdown into multiple languages by setting `charset` attribute to `UTF-8`.
```

**Review by Sboonny — DISMISSED** [review]
I don't think this description is fitting, it's assuming people know what is a meta data.

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md (hunk 3 above)
Comment (hanswang123456) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:10]: sounds good. I've committed.

**Comment by Sboonny** [comment]
the test still accept meta closing tag 

image

image

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md [hunk 4]
```diff
@@ -7,14 +7,19 @@ dashedName: step-68
 
 # --description--
 
-One more thing. You should allow people to use their native language. Tell the browser to encode multiple languages by adding a `meta` element as a child of the `head` element. Set its `charset` attribute to `UTF-8`.
+You can set the browser behavior by using `meta` element. A famous use case is telling the browser to parse the markdown into multiple languages by setting `charset` attribute to `UTF-8`.
```
Comment (moT01) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:10]:
````suggestion
You can set browser behavior by adding self-closing `meta` elements in the `head`. Here's an example:

```
<meta attribute="value" />
```
````

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md [hunk 5]
```diff
@@ -7,14 +7,19 @@ dashedName: step-68
...
+
+Create a `meta` element as a child of the `head` element. Set its `charset` attribute to `UTF-8`.
```
Comment (moT01) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:12]:
```suggestion
Tell the browser to parse the markdown into multiple languages by creating a `meta` element as a child of the `head` element. Set its `charset` attribute to `UTF-8`.
```

**Review by moT01 — COMMENTED** [review]
I made a slight suggestion on the wording. It's really just my preference, so we don't have to use it - but I like it a little better.

There's been some discussion on whether or not we should allow using a closing tag to pass. I think that should pass. That's part of my reason for the description suggestion. Providing the example will show them how to create the self-closing tag - and if they still use the closing tag, then just let it pass. That would be my preference.

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md (hunk 4 above)
Comment (bbsmooth) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:10]: > and if they still use the closing tag, then just let it pass.

I'm not saying this would be the end of the world, but I don't think the HTML specs allow for a closing `</meta>` tag and the W3 HTML validator throws an error if you have one. My understanding is that we want to allow valid HTML wherever possible. Conversely, I think we should fail invalid HTML as well.

Comment (naomi-lgbt) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:10]: We've had a lot of back and forth on the `/>` for self-closing elements. Pretty sure we don't need the `/` and shouldn't include in the example?

__
[sema-logo](https://semasoftware.com/gh) &nbsp;**Summary:** :question: I have a question

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md [hunk 6]
```diff
@@ -7,14 +7,19 @@ dashedName: step-68
...
 # --hints--
 
-You should create a `meta` element within the `head` element.
+You should create a self-closing `meta` element within the `head` element.
 
 ```js
 assert.exists(document.querySelector('head > meta'));
+let innerText = document.querySelector(" head ").innerHTML;
+let matched = (innerText.match(/meta/g) || []).length
+assert.equal(matched, 1)
```
Comment (Sboonny) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:22]: Hi @hanswang123456, the current test doesn't stop `<meta></meta>`

If you need help solving this, join us in [our chat room](https://discord.gg/PRyKn3Vbay) or [the forum](https://forum.freecodecamp.org/c/contributors/3)

Comment (hanswang123456) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:22]: Sounds good, Ill join the chatroom

Comment (raisedadead) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:22]: @hanswang123456 are you still stuck?

Comment (hanswang123456) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:22]: Sorry for the late response, a challenge I've been running into is detecting the closing meta tag's existence. When users include the tag, it is ignored by the browser. I'll need to use another method other than innerHTML(which does not return the closing meta tag)

Comment (raisedadead) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:22]: No worries - keep up the excellent work.

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md (hunk 4 above)
Comment (moT01) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:10]: > Pretty sure we don't need the / and shouldn't include in the example?

I'm fine with either method here.

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md (hunk 6 above)
Comment (moT01) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:22]: From other review thread:

> I don't think the HTML specs allow for a closing </meta> tag and the W3 HTML validator throws an error if you have one. My understanding is that we want to allow valid HTML wherever possible. Conversely, I think we should fail invalid HTML as well.

While the validator may fail with this, the browser still renders the original meta tag as expected and just ignores the closing tag. Typically, I might say that we don't need to test things like this because the browser gives us what we expect and we can't test every edge case. But this is the first project, I would be fine with a test for it. 

I would add a third test for it @hanswang123456, the first test can be reverted to what it was (just check that the meta tag exists). The second one can stay as it is. And the third test can be `You should not have a closing meta tag` or `You should not have a closing </meta> tag`. The test will have to just use regex since nothing is rendered if a closing tag is added.

Comment (hanswang123456) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:22]: > From other review thread:
> 
> > I don't think the HTML specs allow for a closing  tag and the W3 HTML validator throws an error if you have one. My understanding is that we want to allow valid HTML wherever possible. Conversely, I think we should fail invalid HTML as well.
> 
> While the validator may fail with this, the browser still renders the original meta tag as expected and just ignores the closing tag. Typically, I might say that we don't need to test things like this because the browser gives us what we expect and we can't test every edge case. But this is the first project, I would be fine with a test for it.
> 
> I would add a third test for it @hanswang123456, the first test can be reverted to what it was (just check that the meta tag exists). The second one can stay as it is. And the third test can be `You should not have a closing meta tag` or `You should not have a closing </meta> tag`. The test will have to just use regex since nothing is rendered if a closing tag is added.

Sounds good. I'll update the 3rd test .

Comment (Sboonny) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md]: Thank you @hanswang123456, I have update the test to close the issue. Thank you for the hard work and bearing with us

**Comment by hanswang123456** [comment]
@Sboonny the PR seems to be coming together. Are there any other things to fix for the description?

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md [hunk 7]
```diff
@@ -7,11 +7,17 @@ dashedName: step-69
 
 # --description--
 
-One more thing. You should allow people to use their native language. Tell the browser to encode multiple languages by adding a `meta` element as a child of the `head` element. Set its `charset` attribute to `UTF-8`.
+You can set browser behavior by adding self-closing `meta` elements in the `head`. Here's an example:
+
+```
```
Comment (Sboonny) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:12]: add `html` language

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md [hunk 8]
```diff
@@ -7,11 +7,17 @@ dashedName: step-69
...
+<meta attribute="value" />
```
Comment (Sboonny) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:13]: Remove `/` as mentioned in the discussion

**Review by Sboonny — CHANGES_REQUESTED** [review]
to please linter, and following Naomi point https://github.com/freeCodeCamp/freeCodeCamp/pull/47221#discussion_r965245096

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md (hunk 8 above)
Comment (naomi-lgbt) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:13]:
```suggestion
<meta attribute="value">
```

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md (hunk 7 above)
Comment (hanswang123456) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md]: I've updated the PR

**Review by Sboonny — DISMISSED** [review]

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md [hunk 9]
```diff
@@ -23,6 +29,12 @@ You should give the `meta` element a `charset` of `UTF-8`.
 assert.equal(document.querySelector('head > meta')?.getAttribute('charset')?.toLowerCase(), 'utf-8');
 ```
 
+Your `meta` element should be a self-closing tag, you don't need to add `</meta>`.
+
+```js
+assert(code.match(/<\/meta\s*>?/i) === null);
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md:35]:
```suggestion
assert.notMatch(code, /<\/meta\s*>?/i);
```

We can use the specific assert API method for this.

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

**Review by naomi-lgbt — APPROVED** [review]

**Review by Sboonny — APPROVED** [review]
